### PR TITLE
Feature/applied rules

### PR DESCRIPTION
**Pull request by Babarkamal14**
https://trello.com/c/kpVrv5q1/2034-ui-show-applied-alert-rules-list-with-devices-names-on-which-rule-is-applied

### CODE REVIEW

File: src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js [hunk 1]
```diff
@@ -1,17 +1,20 @@
...
 import { getLimitText } from '../rulesEngine.utils';
+import { getGatewayById } from '../../inventory/gateways/gateways.utils';
 
 class AddRule extends Component {
   constructor(props) {
     super(props);
     this.state = {
       isAdd: false,
       ruleId: '',
+      visible: false,
```
Comment (hasansiddique-wanclouds) [inline src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js]: change it to `showConfirmPopup`...

File: src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js [hunk 2]
```diff
@@ -1,17 +1,20 @@
...
+      currentGateway: '',
```
Comment (hasansiddique-wanclouds) [inline src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js]: change it to `selectedDeleteGateway` as it only used for selecting delete gateway...

File: src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js [hunk 3]
```diff
@@ -21,14 +24,64 @@ class AddRule extends Component {
...
+        >
+          <strong>Device: </strong>{gateway.name}
+          {this.state.currentGateway === gateway.id && this.props.isDeleting && <Spin size="small" />}
+        </Tag>
+      </Popconfirm>
+    );
+  };
+
+  togglePopconfirm = (e, gateway) => {
+    e.preventDefault();
```
Comment (hasansiddique-wanclouds) [inline src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js]: `e` should be change to `event` to remove any ambiguity.

File: src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js [hunk 4]
```diff
@@ -21,14 +24,64 @@ class AddRule extends Component {
...
+    });
+  };
+
+  handleCancel = () => {
+    this.setState({
+      visible: false,
+    });
+  };
+
+  deleteGateway = (gatewayId, ruleId) => {
```
Comment (hasansiddique-wanclouds) [inline src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js]: rename it to `handleRuleGatewayDelete`.

File: src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js [hunk 5]
```diff
@@ -88,22 +141,40 @@ class AddRule extends Component {
...
-                                (rule.lessThan || rule.greaterThan || rule.fixedValue)}
-                              </Tag> }
-                            <Tag><strong>Duration: </strong> {rule.time}</Tag>
-                          </Col>
-                        </Row>
-                      }
-                    />
+                    <Row>
+                      <Col span={24}>
+                        <List.Item.Meta
```
Comment (hasansiddique-wanclouds) [inline src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js]: I dont thin we need `Row` and `Col` tags here. Remove them.

File: src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js [hunk 6]
```diff
@@ -88,22 +141,40 @@ class AddRule extends Component {
...
+                          title={startCase(rule.ruleType)}
+                          description={
+                            <Row gutter={16}>
+                              <Col span={24} className="gutter-row">
+                                {includes(rule.ruleType, 'status') && rule.fixedValue ?
```
Comment (hasansiddique-wanclouds) [inline src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js]: I dont think includes is used properly. It should be compared with some array.

Comment (hasansiddique-wanclouds) [reply src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js]: If it is about checking that the ruleStatus exist then change it to `get`...

File: src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js [hunk 7]
```diff
@@ -88,22 +141,40 @@ class AddRule extends Component {
...
+                                    {getLimitText(rule) +
+                                    (rule.lessThan || rule.greaterThan || rule.fixedValue)}
+                                  </Tag> }
+                                <Tag><strong>Duration: </strong> {rule.time}</Tag>
+                              </Col>
+                            </Row>
+                          }
+                        />
+                      </Col>
+                      {rule.gateways.length !== 0 &&
```
Comment (hasansiddique-wanclouds) [inline src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js]: It should be `size` not length as we are using immutable lists...

File: src/dashboard/layout/content/rulesEngine/rulesEngine.api.js [hunk 8]
```diff
@@ -136,3 +139,33 @@ export const deleteRule = (ruleId) => {
     });
   };
 };
+export const deleteAppliedRule = (data) => {
```
Comment (hasansiddique-wanclouds) [inline src/dashboard/layout/content/rulesEngine/rulesEngine.api.js]: change this function name as it is not for deleting rule rather APPLYING/DISAPPLYING rule.
change it to `applyDisapplyRulesOnGateways`.

File: src/dashboard/layout/content/rulesEngine/rulesEngine.reducer.js [hunk 9]
```diff
@@ -92,6 +99,25 @@ const rulesEngine = (state = initialState, action) => {
         .set('isDeleting', false);
     default:
       return state;
+
+    case DELETE_APPLIED_RULE_REQUEST:
+      return state
+        .set('isDeleting', true);
+
+    case DELETE_APPLIED_RULE_SUCCESS:
+      index = state.get('rules').findIndex((rule) => {
```
Comment (hasansiddique-wanclouds) [inline src/dashboard/layout/content/rulesEngine/rulesEngine.reducer.js]: `index` is used for integer index, you are using it as object which is logically wrong.
Update you logic.

File: src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js (hunk 1 above)
Comment (Babarkamal14) [reply src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js]: done

File: src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js (hunk 2 above)
Comment (Babarkamal14) [reply src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js]: done

File: src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js (hunk 3 above)
Comment (Babarkamal14) [reply src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js]: done

File: src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js (hunk 4 above)
Comment (Babarkamal14) [reply src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js]: done

File: src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js (hunk 7 above)
Comment (Babarkamal14) [reply src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js]: done

File: src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js (hunk 5 above)
Comment (Babarkamal14) [reply src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js]: done

File: src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js (hunk 4 above)
Comment (Babarkamal14) [reply src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js]: done

File: src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js (hunk 2 above)
Comment (Babarkamal14) [reply src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js]: done

File: src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js (hunk 1 above)
Comment (Babarkamal14) [reply src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js]: done

File: src/dashboard/layout/content/rulesEngine/rulesEngine.api.js (hunk 8 above)
Comment (Babarkamal14) [reply src/dashboard/layout/content/rulesEngine/rulesEngine.api.js]: done

File: src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js [hunk 10]
```diff
@@ -21,14 +24,64 @@ class AddRule extends Component {
     }
   }
 
+  rulesAppliedOn = (gatewayId, ruleId) => {
+    const gateway = getGatewayById(gatewayId);
+    return (
+      <Popconfirm
+        placement="topRight"
+        title="Are you sure you want to remove this device?"
```
Comment (hasansiddique-wanclouds) [inline src/dashboard/layout/content/rulesEngine/addRule/AddRule.view.js]: use resourcestrings for these messages.

**Pull request merged**